README.md
# `chonk`

[![Greenkeeper badge](https://badges.greenkeeper.io/decompil3d/chonk.svg)](https://greenkeeper.io/)

Check the size of a file and determine its chonkiness

Inspired by <https://knowyourmeme.com/memes/chonk-oh-lawd-he-comin>

## Installation

```sh
# Local
npm i -D chonk

# Global
npm i -g chonk
```

## Usage

```sh
chonk [<options>] <file> <file>...
```

### Options

```sh-session
Options:
  --help       Show help                                               [boolean]
  --version    Show version number                                     [boolean]
  --lawd, -L   Minimum size in kilobytes for OH LAWD, OT COMIN'!
                                                     [number] [default: "50000"]
  --mega, -M   Minimum size in kilobytes for MEGACHONKER
                                                     [number] [default: "10000"]
  --hefty, -H  Minimum size in kilobytes for HEFTYCHONK
                                                      [number] [default: "2000"]
  --heck, -h   Minimum size in kilobytes for A heckin' chonker
                                                       [number] [default: "500"]
  --chonk, -c  Minimum size in kilobytes for It chonk   [number] [default: "50"]
```

### Sample

```sh-session
$ chonk lawd.dmg mega.zip hefty.tar.gz heckin.jpg chonk.pdf fine.txt
```

![sample output](chonk.png)


CHANGELOG.md
# `chonk`

## 2.0.1

- Update dependencies

## 2.0.0

- Drop Node.js 10 support
- Update deps

## 1.0.4

- Update dependencies

## 1.0.3

- Update more dependencies

## 1.0.2

- Update dependencies

## 1.0.1

- Tests

## 1.0.0

- Initial release



test/unit/index.test.js
const assume = require('assume');
const { describe, it } = require('mocha');
const { getChonkiness } = require('../../index');

describe('getChonkiness', function () {
  global.argv = {
    lawd: 50000,
    mega: 10000,
    hefty: 2000,
    heck: 500,
    chonk: 50
  };

  it('is a function', function () {
    assume(getChonkiness).is.a('function');
    assume(getChonkiness).has.length(1);
  });

  function assumeSizeResponse(size, expectedResponse) {
    return function _assumeSizeResponse() {
      const res = getChonkiness(size);
      assume(res.text).equals(expectedResponse);
    };
  }

  it('correctly identifies a fine file', assumeSizeResponse(5000, 'A fine file'));
  it('correctly identifies a chonker', assumeSizeResponse(50001, 'It chonk'));
  it('correctly identifies a a heckin\' chonker', assumeSizeResponse(500001, 'A heckin\' chonker'));
  it('correctly identifies a heftychonk', assumeSizeResponse(2000001, 'HEFTYCHONK'));
  it('correctly identifies a megachonker', assumeSizeResponse(10000001, 'MEGACHONKER'));
  it('correctly identifies when it comin\'', assumeSizeResponse(50000001, 'OH LAWD, IT COMIN\'!'));
});


test/integration/index.test.js
const { spawn } = require('child_process');
const path = require('path');
const assume = require('assume');
const { describe, it } = require('mocha');
const concat = require('concat-stream');

describe('chonk', function () {
  function verifyStd(whichStd, expected) {
    return function (std) {
      if (expected) {
        if (Array.isArray(expected)) {
          let lastFound = -1;
          expected.forEach(o => {
            const idx = std.indexOf(o);
            assume(idx).not.equals(-1, `'${o}' not found in ${whichStd}`);
            assume(idx).is.gt(lastFound, `'${o}' was found, but not in the right order`);
            lastFound = idx;
          });
        } else {
          assume(std).includes(expected);
        }
      }
    };
  }

  function assumeOutput(expectedOut, expectedErr, expectedExitCode, ...args) {
    return function (done) {
      const proc = spawn(path.join(__dirname, '../../index.js'), args);
      const stdoutVerifier = concat({
        encoding: 'string'
      }, verifyStd('stdout', expectedOut));
      const stderrVerifier = concat({
        encoding: 'string'
      }, verifyStd('stderr', expectedErr));
      proc.stdout.pipe(stdoutVerifier);
      proc.stderr.pipe(stderrVerifier);
      proc.on('close', code => {
        assume(code).equals(expectedExitCode);
        done();
      });
    };
  }

  it('errors when no parameters specified', assumeOutput(
    null,
    'Wat? No files specified',
    1
  ));

  it('prints usage', assumeOutput(
    'Options:',
    null,
    0,
    '--help'
  ));

  it('prints files in order', assumeOutput(
    [
      '0.027.txt', 'A fine file',
      '50.txt', 'It chonk',
      '500.txt', 'A heckin\' chonker',
      '2000.txt', 'HEFTYCHONK',
      '10000.txt', 'MEGACHONKER',
      '50001.txt', 'OH LAWD, IT COMIN\'!'
    ],
    null,
    0,
    ...[
      '0.027.txt',
      '50.txt',
      '500.txt',
      '2000.txt',
      '10000.txt',
      '50001.txt'
    ].map(filename => path.join(__dirname, 'fixtures', filename))
  ));

  it('accepts chonkiness override flags', assumeOutput(
    [
      '50.txt', 'A fine file',
      '500.txt', 'It chonk',
      '2000.txt', 'A heckin\' chonker',
      '10000.txt', 'HEFTYCHONK',
      '50001.txt', 'MEGACHONKER'
    ],
    null,
    0,
    ...[
      '-c 51',
      '-h 510',
      '-H 3000',
      '-M 11000',
      '-L 52000'
    ].concat([
      '50.txt',
      '500.txt',
      '2000.txt',
      '10000.txt',
      '50001.txt'
    ].map(filename => path.join(__dirname, 'fixtures', filename)))
  ));

  it('bails on invalid path', assumeOutput(
    null,
    'Error:',
    1,
    'someNonExistantFile.txt'
  ));
});


index.js
#!/usr/bin/env node
/* eslint no-console:0, no-process-exit:0 */

const async = require('async');
const chalk = require('chalk');
const fs = require('fs');
const prettyBytes = require('pretty-bytes');
const argv = require('yargs')
  .option('lawd', {
    alias: 'L',
    default: '50000',
    describe: 'Minimum size in kilobytes for OH LAWD, IT COMIN\'!',
    type: 'number'
  })
  .option('mega', {
    alias: 'M',
    default: '10000',
    describe: 'Minimum size in kilobytes for MEGACHONKER',
    type: 'number'
  })
  .option('hefty', {
    alias: 'H',
    default: '2000',
    describe: 'Minimum size in kilobytes for HEFTYCHONK',
    type: 'number'
  })
  .option('heck', {
    alias: 'h',
    default: '500',
    describe: 'Minimum size in kilobytes for A heckin\' chonker',
    type: 'number'
  })
  .option('chonk', {
    alias: 'c',
    default: '50',
    describe: 'Minimum size in kilobytes for It chonk',
    type: 'number'
  })
  .argv;

if (!argv._ || argv._.length < 1) {
  console.error(chalk.red('Wat? No files specified 😿'));
  process.exit(1);
}

async.eachSeries(argv._, (filePath, next) => {
  if (typeof filePath !== 'string') return next(new Error('invalid file path'));
  fs.stat(filePath, (statErr, stats) => {
    if (statErr) return next(statErr);

    const { size } = stats;
    const chonkiness = getChonkiness(size);
    console.log(chalk.white(filePath + ':'), chalk[chonkiness.color](chonkiness.text), `(${prettyBytes(size)})`);
    next();
  });
}, err => {
  if (err) {
    console.error(chalk.red('Error:'), err.message);
    process.exit(1);
  }
});

const KB = 1000;

function getChonkiness(size) {
  // @ts-ignore
  if (size > argv.lawd * KB) {
    return {
      color: 'red',
      text: 'OH LAWD, IT COMIN\'!'
    };
  }
  // @ts-ignore
  if (size > argv.mega * KB) {
    return {
      color: 'red',
      text: 'MEGACHONKER'
    };
  }
  // @ts-ignore
  if (size > argv.hefty * KB) {
    return {
      color: 'yellow',
      text: 'HEFTYCHONK'
    };
  }
  // @ts-ignore
  if (size > argv.heck * KB) {
    return {
      color: 'yellow',
      text: 'A heckin\' chonker'
    };
  }
  // @ts-ignore
  if (size > argv.chonk * KB) {
    return {
      color: 'cyan',
      text: 'It chonk'
    };
  }
  return {
    color: 'green',
    text: 'A fine file'
  };
}

// Exports for testing
module.exports = {
  getChonkiness
};
